cmd/juju/application/showunit.go

```diff
@@ -0,0 +1,283 @@
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const showUnitDoc = `
+The command takes deployed unit names as an argument.
+
+Optionally, relation data for only a specified endpoint
+or related unit may be shown, or just the application data. 
+
+Examples:
+    $ juju show-unit mysql/0
+    $ juju show-unit mysql/0 wordpress/1
+    $ juju show-unit mysql/0 --app
+    $ juju show-unit mysql/0 --endpoint db
+    $ juju show-unit mysql/0 --related-unit wordpress/2
+`
+
+// NewShowUnitCommand returns a command that displays unit info.
+func NewShowUnitCommand() cmd.Command {
+	s := &showUnitCommand{}
+	s.newAPIFunc = func() (UnitsInfoAPI, error) {
+		return s.newUnitAPI()
+	}
+	return modelcmd.Wrap(s)
+}
+
+type showUnitCommand struct {
+	modelcmd.ModelCommandBase
+
+	out         cmd.Output
+	units       []string
+	endpoint    string
+	relatedUnit string
+	appOnly     bool
+
+	newAPIFunc func() (UnitsInfoAPI, error)
+}
+
+// Info implements Command.Info.
+func (c *showUnitCommand) Info() *cmd.Info {
+	showCmd := &cmd.Info{
+		Name:    "show-unit",
+		Args:    "<unit name>",
+		Purpose: "Displays information about a unit.",
+		Doc:     showUnitDoc,
+	}
+	return jujucmd.Info(showCmd)
+}
+
+// Init implements Command.Init.
+func (c *showUnitCommand) Init(args []string) error {
+	if len(args) < 1 {
+		return errors.Errorf("an unit name must be supplied")
+	}
+	c.units = args
+	if c.relatedUnit != "" && !names.IsValidUnit(c.relatedUnit) {
+		return errors.NotValidf("related unit name %v", c.relatedUnit)
+	}
+	var invalid []string
+	for _, one := range c.units {
+		if !names.IsValidUnit(one) {
+			invalid = append(invalid, one)
+		}
+	}
+	if len(invalid) == 0 {
+		return nil
+	}
+	plural := "s"
+	if len(invalid) == 1 {
+		plural = ""
+	}
+	return errors.NotValidf(`unit name%v %v`, plural, strings.Join(invalid, `, `))
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *showUnitCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	c.out.AddFlags(f, "yaml", cmd.DefaultFormatters.Formatters())
+	f.StringVar(&c.endpoint, "endpoint", "", "Only show relation data for the specified endpoint")
+	f.StringVar(&c.relatedUnit, "related-unit", "", "Only show relation data for the specified unit")
+	f.BoolVar(&c.appOnly, "app", false, "Only show application relation data")
+}
+
+// UnitsInfoAPI defines the API methods that show-unit command uses.
+type UnitsInfoAPI interface {
+	Close() error
+	BestAPIVersion() int
+	UnitsInfo([]names.UnitTag) ([]params.UnitInfoResult, error)
+}
+
+func (c *showUnitCommand) newUnitAPI() (UnitsInfoAPI, error) {
+	root, err := c.NewAPIRoot()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return application.NewClient(root), nil
+}
+
+// Info implements Command.Run.
+func (c *showUnitCommand) Run(ctx *cmd.Context) error {
+	client, err := c.newAPIFunc()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	if v := client.BestAPIVersion(); v < 12 {
```

Comment:
Why check this here when it is handled in the api client?

Reply:
We tend to do this quite commonly. It does seem redundant.